File: src/leetCode/linkedList/Problem141.java
package leetCode.linkedList;

/**
 * Given a linked list, determine if it has a cycle in it.
 * <p>
 * Follow up:
 * Can you solve it without using extra space?
 * <p>
 * create by stephen on 2018/5/15
 */
public class Problem141 {
    private class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    /**
     * 快慢指针
     * O(n) time + O(1) space
     */
    public boolean hasCycle(ListNode head) {
        if (head == null || head.next == null) return false;
        ListNode slow = head, fast = head;
        while (fast != null) {
            slow = slow.next;
            fast = fast.next;
            if (fast != null) fast = fast.next;
            if (fast == slow) return true;
        }

        return false;
    }


}

File: leetCode/linkedList/Problem141.cpp
#include <list.h>

/**
 * slow and fast pointers
 * O(n) time + O(1) space
 */
bool hasCycle(ListNode* head) {
  if (head == nullptr || head->next == nullptr) return false;
  ListNode *slow = head, *fast = head;
  while (fast != nullptr) {
    slow = slow->next;
    fast = fast->next;
    if (fast != nullptr) fast = fast->next;
    if (fast == slow) return true;
  }
  return false;
}
